zipline/assets/assets.py

```diff
@@ -678,3 +678,148 @@ class AssetConvertible(with_metaclass(ABCMeta)):
 
 class NotAssetConvertible(ValueError):
     pass
+
+
+class AssetFinderCachedEquities(AssetFinder):
+    """
+    An extension to AssetFinder that loads all equities from equities table
+    into memory and overrides lookup_symbol, which looks the equities up in the
+    in-memory store.
+    """
+    def __init__(self, engine):
+        super(AssetFinderCachedEquities, self).__init__(engine)
+        self.fuzzy_symbol_hashed_equities = {}
+        self.company_share_class_hashed_equities = {}
+        self.hashed_equities = sa.select(self.equities.c).execute().fetchall()
+        self.load_company_share_class_hashed_equities()
+        self.load_fuzzy_symbol_hashed_equities()
+
+    def load_fuzzy_symbol_hashed_equities(self):
+        """
+        Populates map of fuzzy symbol to a list of equities that have that
+        fuzzy symbol.
+        """
+        for equity in self.hashed_equities:
+            fuzzy_symbol = equity['fuzzy_symbol']
+            if fuzzy_symbol not in self.fuzzy_symbol_hashed_equities:
+                self.fuzzy_symbol_hashed_equities[fuzzy_symbol] = []
+            asset = self.convert_row_to_equity(equity)
+            self.fuzzy_symbol_hashed_equities[fuzzy_symbol].append(asset)
+
+    def convert_row_to_equity(self, equity):
+        """
+        Converts a SQLAlchemy equity row to an Equity object.
+        """
+        data = dict(equity.items())
+        _convert_asset_timestamp_fields(data)
+        asset = Equity(**data)
+        return asset
+
+    def load_company_share_class_hashed_equities(self):
+        """
+        Populates map of (company symbol, share class symbol) to a list of
+         equities that have that combination of (company symbol, share class
+         symbol).
+        """
+        for equity in self.hashed_equities:
+            company_symbol = equity['company_symbol']
+            share_class_symbol = equity['share_class_symbol']
+            if (company_symbol, share_class_symbol) not in \
+                    self.company_share_class_hashed_equities:
+                self.company_share_class_hashed_equities[(
+                    company_symbol,
+                    share_class_symbol
+                )] = []
+            asset = self.convert_row_to_equity(equity)
```

Comment:
If this method is called in both caching functions, then are we building each Asset twice? A lot of the logic between caching by fuzzy and caching by company/sc symbol are shared, and they could probably be condensed to one method that iterates through all equities, building each once, and inserting it in to both dicts.